### PR TITLE
refactor(nextly): update a single's schema and row in one place

### PR DESCRIPTION
**Pull request by mobeenabdullah**
## What

Moves `updateSingleSchema`'s DDL body out of the request handler and into
`SingleMetadataService`, which already owns `createSingle` and `deleteSingle`.
The handler drops from 1579 to 1288 lines.

This is the last piece of task 116. It matters because a migration lock cannot be
placed where the schema change and the registry row live in different objects: a lock
taken inside the registry service is acquired *after* the tables have already changed,
so it samples rather than holds.

## How

Re-expressed in the service's existing three-phase contract rather than copied:

| Phase | Contract |
|---|---|
| **PLAN** (`planUpdate`) | Reads the live table, decides create-vs-alter, renders SQL. **May reject** — nothing is persisted and no statement has run. |
| **APPLY** (`applyUpdateDdl`) | Runs statements, confirms the table, registers the runtime schema, reconciles the companion. Returns a status. |
| **RECORD** | `registry.updateSingle`, with the outcome already known. |

The phase boundary replaces the `migrationBegan` flag the handler carried.

### One plan, not two branches

A save that only flips Internationalization or Draft/Published is a plan whose
`migrationSQL` is `""` — not a parallel execution path. Both shapes now share one
apply, one status vocabulary and one failure contract, so they cannot drift.

### 🔴 `instanceof NextlyError` is NOT replaced by the phase split

The handler's catch was two rules:

```js
if (migrationError instanceof NextlyError || !migrationBegan) throw
```

The split replaces `!migrationBegan`. It does **not** replace the first clause, which
is a separate defensive rule: a refusal propagates whatever phase it arrives from.
Without it a future apply-phase refusal would be recorded as `"failed"` and the field
list saved against a table that never took it.

That clause is unreachable today — adapters raise `DatabaseError`, which extends
`Error`, not `NextlyError` — which is what makes carrying it cheap rather than
unnecessary. Unreachability is a property of the current call graph.

### `statusRequested` is passed, not derived

The old condition was `b.status !== undefined`, meaning "the request SET the toggle".
That is not the same as `hasStatus !== wasStatus`: saving Draft/Published at its
current value still reaches the companion, and because provisioning is idempotent,
that save is what repairs a localized single whose companion `_status` was never
created. Deriving it would collapse the repair into a no-op.

## Tests

New file `single-dispatcher-update-ddl.test.ts`. **The update path had no unit
coverage of its DDL at all** — its only unit tests replace the schema service with a
stub returning `""` because they test response envelopes.

Three hazard tests were proven by breaking the code they guard:

| Break | Result |
|---|---|
| Remove the `instanceof NextlyError` rethrow | 1 failed / 6 passed — the apply-phase refusal test, alone |
| Remove the create-vs-alter split | 1 failed / 6 passed — the rebuild test, alone |
| Key the normaliser on NAME instead of COLUMN | **7 passed — did not break** |

That third one is stated honestly in the file rather than shipped as coverage.
Measured directly: both keyings emit byte-identical SQL, because the generator does
not DROP a system column that goes missing from the desired list. The test was
relabelled to say it asserts the outcome and cannot separate the mechanism, and names
`columnsDeclaredBy` as where that choice is decidable.

## Verification

- `check-types --force`, `lint` (by exit code), `check-drizzle-v1-legacy`: all `0`
- Full `packages/nextly` unit run, per-test STATE diff against a freshly measured
  baseline at `9a1525260` (`passed=7416 failed=361 skipped=42 pending=6`):
  **0 moved, 0 removed, 7 added**
- Integration, all three dialects, on the pre-rebase tree: sqlite 1186 passed /
  postgres17 1395 passed / mysql 1311 passed, **0 failed**. Counts confirmed non-zero
  so a self-skipped leg could not read as a pass. CI covers the rebased combination.

### CODE REVIEW

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/nextly/src/domains/singles/services/single-metadata-service.ts [hunk 1]
```diff
@@ -226,6 +299,360 @@ export class SingleMetadataService {
...
+        localized: input.isLocalized,
+        wasLocalized: input.wasLocalized,
+        status: input.hasStatus,
+        wasStatus: input.wasStatus,
+        adapter,
+      });
+
+      return "applied";
+    } catch (error) {
+      if (error instanceof NextlyError) throw error;
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/singles/services/single-metadata-service.ts]: Confirmed, and it changed the design. Fixed in `8d9111f39`.

Traced the path you named: `reconcile-single-companion.ts` executes every statement of its plan, and only then — when `plan.companionDropped` — calls `forgetI18nTransition`, which reaches `markerKey` (`i18n/migration/transition-state.ts:214`) and raises `NextlyError.internal` for a slug containing a dot. So a localization disable can restore, archive and DROP the companion, then raise. The old catch had the same hole (`instanceof NextlyError || !migrationBegan` rethrows on the first clause regardless of phase), so this is pre-existing rather than introduced — but this PR is where the logic was consolidated, so it is the right place to close it.

I have not adopted the `migrationBegan` flag, because the classification does not need one once the phases are separated properly. **The apply now records every error and never consults the type.** Reasoning:

- Everything that can legitimately REJECT runs in the plan, before any statement, where raising costs nothing and nothing is persisted. That is what the phase split is for.
- Once the apply has begun, the schema may already have changed. Raising there skips the registry write, and a row that never learns what happened leaves the registry describing storage that no longer matches it — silent, persistent, and with nothing recording the state. `failed` is visible, accurate and retryable.

So a boolean tracking "has schema work begun" would only re-derive a boundary the phase split already draws; the rule that fits is simply that the apply reports rather than raises.

You are also right that the old test did not distinguish the case: it threw on the first statement, which is the same side of the boundary. Replaced with one that injects the failure AFTER the companion reconcile has completed its DDL, on a localization disable, and asserts the row records `failed`. Proven load-bearing — reinstating `if (error instanceof NextlyError) throw error` fails exactly that test and nothing else (1 failed / 6 passed).

File: packages/nextly/src/dispatcher/handlers/__tests__/single-dispatcher-update-ddl.test.ts [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+  const registry = wireRegistry(options.existing ?? existingSingle());
+  const result = await dispatchSingles(
+    "updateSingleSchema",
+    { slug: "page" },
+    payload
+  );
+  const written = registry.updateSingle.mock.calls[0]?.[1] as
+    | { migrationStatus?: string }
+    | undefined;
+  return { registry, result, sql: executed.join("\n"), written };
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/dispatcher/handlers/__tests__/single-dispatcher-update-ddl.test.ts:282]: Confirmed, and there was a product defect under it. Fixed in `8d9111f39`.

You are right that `written` is only what the double was handed. Checking what the real registry does with it, `SingleRegistryService.updateSingle` wrote `migration_status` **only inside `if (data.fields)` and only when `fieldsActuallyChanged || statusToggled`**. Two consequences, both real:

1. **The rebuild path could never record success.** A single whose create failed carries `migration_status: "failed"`. Re-saving it sends the SAME fields, so `schemaHashesMatch` is true, `fieldsActuallyChanged` is false, the gate is closed — and the one save able to repair it could not record that it had. It stays `failed` however many times it is retried. Exactly as you predicted for the missing-table retry.
2. **A flag-only save carries no `fields` at all**, so it never reached the gate and its companion provisioning went unrecorded whatever the outcome.

Rather than make the double faithful — which would be a second implementation of the gate, and would drift — I fixed the gate and tested it where it lives:

```ts
if (data.migrationStatus !== undefined) {
  updateData.migration_status = data.migrationStatus;
}
```

The distinction is between a GUESS and an OBSERVATION. `"pending"` is a guess about work still to do, and suppressing it on a metadata-only sync is exactly what the gate should keep doing — that purpose is untouched, and `schema_version` still bumps only on a physical change. A status the caller passes explicitly is not a guess: `SingleMetadataService` has run the statements and confirmed the table before it hands one over, and the registry is the only place that answer can be kept.

Three tests added to `single-registry-service.status-persist.test.ts`, which drives the REAL service and asserts the row the adapter is actually handed:

- an explicit status is recorded when the fields did not change (and `schema_version` still does not bump)
- an explicit status is recorded when the save carries no fields
- **control:** with no status stated, `migration_status` is still absent from the update

Both directions proven by breaking the code: suppressing the new write fails the first two and leaves the control green; writing unconditionally fails only the control. Full `packages/nextly` unit run against a freshly measured baseline at `68145f1ab`: 0 moved, 0 removed, 10 added — no existing test moved, which was the thing to check when loosening a shared gate.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/nextly/src/domains/singles/services/single-registry-service.ts [hunk 3]
```diff
@@ -493,6 +493,23 @@ export class SingleRegistryService extends BaseRegistryService<
...
+    // status the caller passes explicitly is not a guess: `SingleMetadataService` has just run the
+    // statements and confirmed the table, and this is the only place that answer can be kept.
+    //
+    // Suppressing it left two saves unable to record what they had already done. A flag-only save
+    // carries no `fields`, so it never reached the gate at all and its companion provisioning went
+    // unrecorded. And a re-save of UNCHANGED fields against a single whose create failed is exactly
+    // the rebuild path — same schema hash, so `fieldsActuallyChanged` is false — which meant the
+    // one save able to repair a `failed` single could never record that it had.
+    if (data.migrationStatus !== undefined) {
+      updateData.migration_status = data.migrationStatus;
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/singles/services/single-registry-service.ts:510]: Confirmed, and it was a regression my previous round introduced. Fixed in `4b8bec25e`.

Before that round a partially-created table stayed `failed`, which was at least honest. Making the registry record an explicitly-supplied status turned that into an unconditional upgrade, and you are right that confirming the MAIN TABLE EXISTS is not confirming the schema: a create whose `CREATE TABLE` succeeded and whose index or junction-table statement failed leaves the table present and incomplete, and an unchanged-field re-save takes the alter branch and emits nothing.

The apply now refuses to clear a durable `failed` when it ran no statements:

```ts
if (statementsRun === 0 && input.existing.migrationStatus === "failed") {
  return "failed";
}
```

🔴 **The obvious form of that guard silently never fires, which is worth recording.** I first wrote it as `!plan.migrationSQL` — and a diff with no operations still renders a header comment (`-- Update dynamic collection: single_page`), so the string is non-empty while nothing executes. The test caught it; a review might not have.

So the count comes from the runner rather than from the text. `applyMigrationStatements` now returns how many statements it dispatched, because that is where the splitting rule lives and any caller re-deriving it from the string would be a second implementation of that rule. Its own suite gained a case pinning that a comment-only migration counts zero, and the two existing cases now assert the count instead of `toBeUndefined()` — which strengthens them, since asserting `2` for the MySQL index case makes "kept going past the tolerated statement" observable from the return value and not only from `executed`.

Three tests, each proven by breaking the code: claiming `applied` regardless fails the durable-verdict test; returning `max(dispatched, 1)` for a non-empty string fails both the runner count test and the service test that depends on it.

**What I deliberately have NOT done, and why:** actually repairing the missing artifacts by re-running the idempotent create plan for a `failed` row. `applyMigrationStatements` already tolerates "already exists", so the machinery is there, but choosing between the create plan and the alter for a row that may ALSO have field changes is a real design question and needs cross-dialect proof. Filed as its own task. This PR stops the overclaim; it does not yet perform the repair, and the status now says so rather than implying otherwise.

File: packages/nextly/src/domains/singles/services/single-metadata-service.ts [hunk 4]
```diff
@@ -226,6 +299,374 @@ export class SingleMetadataService {
...
+      // Observed, not assumed, and it covers the rebuild case as well as the alter: a plan that
+      // created the table has to find it afterwards for "applied" to mean anything.
+      if (!(await adapter.tableExists(tableName))) {
+        this.logger.error(
+          `[Singles] Table "${tableName}" not found after migration update`
+        );
+        return "failed";
+      }
+
+      await this.registerUpdatedRuntimeSchema(input, plan, adapter);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/singles/services/single-metadata-service.ts]: Confirmed, and it is a regression I introduced by unifying the two branches. Fixed in `4b8bec25e`.

Checked the pre-relocation handler: the fields branch registered the runtime shape BEFORE `reconcileSingleCompanion`, while the flag-only branch reconciled FIRST and registered afterwards. Unifying them onto one apply adopted the fields-branch order for both, so the flag-only localization enable lost the safe ordering it had.

Your description of the consequence checks out at the seam. `ensure-runtime-table.ts` records only the shapes IT registered, and states plainly that "no record means the registration came from boot / create-time / the reconcile, which own it" — so a registration this apply made is ADOPTED rather than rebuilt. Bind a `localized: true` main shape, have the companion CREATE fail, and the resolver describes a main table without translatable columns that are still physically there, until a restart.

Registration now happens after the reconcile has returned, so a failed reconcile leaves the previous shape in place. That is the safe direction in both cases: the previous shape is at worst stale in the way it already was before the save, and the lazy rebuild can still correct it, whereas a shape derived from a state the apply never reached is authoritative and wrong.

Two tests: with the companion failing, `registerDynamicSchema` must not be called; and a control on the success path asserting it IS called exactly once — without that, an apply which stopped registering altogether would satisfy the first while silently dropping the rebinding every successful save depends on. The adapter double now carries a `tableResolver` so the call is observable at all; the create-side sibling omits it on purpose, which is why this needed its own.

Proven by breaking: moving the registration back above the reconcile fails exactly the first test and leaves the control green.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/nextly/src/domains/singles/services/single-metadata-service.ts [hunk 5]
```diff
@@ -226,6 +299,408 @@ export class SingleMetadataService {
...
+      // Leaving the previous shape in place on failure is the safe direction: it is at worst stale
+      // in the same way it was before the save, and the lazy rebuild can still correct it.
+      await this.registerUpdatedRuntimeSchema(input, plan, adapter);
+
+      // A plan that rendered no main-table statements never touched the main table, so it cannot
+      // vouch for one. A create that got its `CREATE TABLE` through and then failed on an index or
+      // a junction table leaves the table PRESENT but incomplete; re-saving unchanged fields takes
+      // the alter branch, emits nothing, and finds the table there. Confirming existence is not
+      // confirming the schema, so the durable verdict stands until something actually repairs it.
+      if (statementsRun === 0 && input.existing.migrationStatus === "failed") {
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/singles/services/single-metadata-service.ts]: Right again, and the rule was wrong rather than merely incomplete. Fixed in `d91073fc2`.

`statementsRun > 0` proves something ran; it does not prove the RIGHT thing ran. An unrelated field edit against the same incomplete table emits a real ALTER, runs plenty of statements, and goes nowhere near the index that never got created — so counting statements clears the one durable record that anything is wrong.

The rule now asks what the plan **describes**, not what it executed:

```ts
if (input.existing.migrationStatus === "failed" && !plan.describesWholeTable) {
  return "failed";
}
```

`describesWholeTable` is true only on the rebuild branch, where the table is absent and the plan renders it from the desired spec in full — every column, index and junction table. An ALTER describes a delta and re-establishes nothing it does not mention, so it can never vouch for a table whose create failed part way. That is the property that separates the two, and it is a fact about the plan rather than about the run, so it cannot be satisfied accidentally by volume.

Three tests, each proven by breaking the code:
- a `failed` single whose save emits no statements stays failed
- **a `failed` single whose ALTER really does run stays failed** — your case; the assertion checks the ALTER was emitted, so the test cannot pass by the alter never happening
- **control:** a `failed` single WITH its table absent is cleared to `applied`, because the rebuild does re-establish the whole schema

Reverting to `!plan.migrationSQL` fails the first two; making the guard unconditional fails only the control.

**I also reverted the statement count I added last round.** With the rule keyed on the plan, `applyMigrationStatements` has no caller for it, and shipping an unused return value on a shared runner is not worth the surface. It is back to `Promise<void>`, and its two tests are back to `toBeUndefined()`. I kept the case I added — a comment-only migration runs nothing — since that asymmetry is exactly what made "the SQL is not empty" the wrong question, and it is worth pinning where the renderer output is produced.

Unit against a freshly measured baseline at `68145f1ab`: 0 moved, 0 removed, 16 added. Integration on all three dialects running.

The actual repair — recreating the missing artifact — remains deliberately out of scope and is filed. This PR now refuses to claim anything it has not established, which is the part that was wrong.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/nextly/src/domains/singles/services/single-metadata-service.ts [hunk 6]
```diff
@@ -226,6 +308,414 @@ export class SingleMetadataService {
...
+      // localization ENABLE it omits the translatable columns — which are still physically on the
+      // main table until the companion has taken them. Bind it first and a companion CREATE that
+      // then fails leaves the resolver describing a main table that does not exist in that shape,
+      // and `ensureSingleRuntimeTable` treats a registration it did not make as owned by whoever
+      // made it, so it adopts this one instead of rebuilding. Reads and writes would drop those
+      // fields until a restart.
+      //
+      // Leaving the previous shape in place on failure is the safe direction: it is at worst stale
+      // in the same way it was before the save, and the lazy rebuild can still correct it.
+      await this.registerUpdatedRuntimeSchema(input, plan, adapter);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/singles/services/single-metadata-service.ts]: Confirmed — this is the flip side of the fix I made for your previous finding, and it showed that BOTH orderings are wrong in one direction. Fixed in `5c5ebdfa7`.

Registering before the reconcile binds a shape whose translatable columns have not moved yet. Registering after it, as I had it, leaves a stale shape when the main ALTER succeeded and the companion did not — your case exactly. Ordering alone cannot fix this, because the shape has TWO independent inputs and a partial apply leaves them in different states:

- the FIELD SET, changed by the main ALTER
- where the translatable columns physically live, changed by the companion reconcile

So the shape that is true after a failed reconcile is **the new field set under the PREVIOUS localization**, which is neither end whole.

I first looked for the option that avoids guessing at all — invalidating the resolver entry so `ensureSingleRuntimeTable` rebuilds it. `SchemaRegistry` has no per-table removal, only `clear()`, which wipes every table. Adding that API is a wider change than this PR should carry, and it would not help anyway: the entry that needs replacing is the one whose FIELD SET is stale, and rebuilding lazily from the row would produce the desired-state shape, which is the wrong one here.

So the reconcile now has its own `try`, and the failure path binds `plan.fields` under `input.wasLocalized`:

```ts
try {
  await reconcileSingleCompanion({ ... });
} catch (companionError) {
  await this.registerUpdatedRuntimeSchema(input, plan, adapter, input.wasLocalized);
  ...
  return "failed";
}
await this.registerUpdatedRuntimeSchema(input, plan, adapter, input.isLocalized);
```

No flag tracks how far the apply got: that inner `catch` is reachable only once the main statements and the existence check have passed, so **the position in the method is the evidence**. `registerUpdatedRuntimeSchema` now takes the localization state as an argument rather than reading `input.isLocalized`, so the caller states what it OBSERVED instead of what the save intended.

Two tests, both asserting the COLUMNS of the table actually bound rather than merely that binding happened:

- companion fails mid-enable → the bound shape still contains `heading`, because a text field translates by default and the enable never moved it off main
- **your case:** an already-localized single adds a non-translatable `count`, the ALTER succeeds, the companion fails → the bound shape contains `count`

Proven by breaking: binding `input.isLocalized` on the failure path fails the first; binding nothing — the state this PR was in one commit ago — fails both.

Full unit against a freshly measured baseline at `175ed53cc`: 0 moved, 0 removed, 17 added.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/nextly/src/domains/singles/services/single-metadata-service.ts [hunk 7]
```diff
@@ -226,6 +308,455 @@ export class SingleMetadataService {
...
+        //
+        // Something has to be bound rather than nothing, because `ensure-runtime-table.ts` adopts a
+        // registration it did not make instead of rebuilding, and the registry has no way to
+        // invalidate one table — so a stale entry survives until a restart either way.
+        await this.registerUpdatedRuntimeSchema(
+          input,
+          plan,
+          adapter,
+          input.wasLocalized
+        );
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/singles/services/single-metadata-service.ts]: Confirmed on both counts, including the criticism of my test. Fixed in `4e5bd76c0`.

You are right that the double threw only after the real reconcile returned while the assertion described the pre-transition shape — so the fixture never reached the state it claimed to be testing. It passed for the wrong reason.

More importantly, this is the THIRD composition to be wrong, and that is the actual finding. Binding the desired shape breaks an enable that never moved the columns; binding nothing breaks a main ALTER that added one; binding the previous shape breaks a disable that restored the columns and failed afterwards. Each is right for one stopping point inside the reconcile and wrong for another, and **the stopping point is not observable from outside it.** Picking a fourth composition would just relocate the same bug.

So the failure path no longer binds anything. It RETRACTS:

```ts
this.retractRuntimeSchema(adapter, tableName);
```

`SchemaRegistry` gained `retractDynamicSchema(tableName)` — deleting the dynamic entry and its edges, and nulling `relationsCache` for the same reason registration does, since relations close over table objects. That was the missing counterpart: a caller could bind a shape but never say "I no longer know what this is", which is why every failure path had to guess and why a wrong guess survived to the next restart.

**Retraction is not a fourth guess, it routes to the authority.** `ensure-runtime-table.ts` rebuilds when the entry is missing, and its rebuild branch already asks `resolveCompanionReadiness` where the translatable columns PHYSICALLY live — the exact fact all three compositions were standing in for. The mechanism was already there; nothing could reach it because nothing could retract.

The double now injects on either side of the real reconcile (`when: "before" | "after"`), which is what your point about separating the states requires, and the recovery is asserted for both:
- `before` — stopped at its first statement, columns never moved
- `after` — completed its DDL and failed on the tail, which is the dotted-slug disable

Both must retract and must NOT register. Proven by breaking: rebinding `wasLocalized` fails both cases; retracting unconditionally fails the success control, which now also asserts no retraction happens on the good path.

Registry-level tests too: retracting forgets one table and leaves its neighbours, leaves a STATIC table of the same name reachable (removing the override, not the name — the failure a blanket `clear()` causes and the reason this is a separate operation), and is a no-op for a name never registered.

Unit against a freshly measured baseline at `175ed53cc`: 0 moved, 0 removed, 20 added.

File: packages/nextly/src/domains/singles/services/single-metadata-service.ts [hunk 8]
```diff
@@ -226,6 +308,455 @@ export class SingleMetadataService {
...
+      // unrelated field edit can run plenty of statements without going anywhere near the missing
+      // artifact. Neither "the table exists" nor "something ran" is evidence the schema is whole,
+      // and the durable verdict is the only record that it is not.
+      //
+      // The rebuild branch is the exception because it renders the table from the desired spec in
+      // full, so reaching the end of it does mean every artifact was asked for.
+      if (
+        input.existing.migrationStatus === "failed" &&
+        !plan.describesWholeTable
+      ) {
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/singles/services/single-metadata-service.ts:694]: Confirmed as a real limitation, and I am accepting it deliberately rather than fixing it here. Reasoning, since a deferral is only honest if the cost is stated.

The case is exact: a localized create builds the main table, fails provisioning the companion, and records `failed`. Every retry then finds the main table present, so the plan is an ALTER with `describesWholeTable: false` — and a retry whose reconcile SUCCEEDS still cannot clear the row. The schema is complete and the status says otherwise, permanently.

What would be needed to fix it properly is knowledge the row does not carry. `migration_status` is one bit: it records THAT a migration failed, not WHICH half. Clearing it correctly means establishing that both halves are now right, and the three narrower options each fail:

- **Infer from the phase that just succeeded** — a successful reconcile says nothing about the main table, which is where the earlier failure might equally have been.
- **Track the failing phase in the row** — a schema change to `dynamic_singles` and a migration, which is a wider change than this PR should carry under its title, and it still only records the phase of the LAST failure.
- **Assume a successful apply implies completeness** — the assumption this PR has already been wrong about twice, in exactly the direction that made a partially created table read as `applied`.

The right answer is to VERIFY rather than infer: compare the live table to the desired spec, indexes and junction tables included, and clear `failed` when it matches. That is post-apply verification, which is already a queued decision here and subsumes both this and the missing-artifact case in
`2026-08-12-0130-failed-single-is-detected-not-repaired.md` — I have folded this scenario into that task rather than opening a third.

The choice between the two costs is what decides it. Wrongly reporting `applied` hides an incomplete schema, and the user finds out when a uniqueness constraint does not hold or a relationship has no junction table to write through. Wrongly holding `failed` is visible, is accurate about the last thing that actually failed, and blocks nothing — reads and writes work against whatever the table is. Given this PR has already shipped the first kind of error twice, holding the conservative position until something can verify is the right side to be wrong on.

The limitation is now stated in the guard comment rather than left for the next reader to discover.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/nextly/src/dispatcher/handlers/__tests__/single-dispatcher-update-ddl.test.ts [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+  "../../../domains/singles/services/reconcile-single-companion",
+  async () => {
+    const actual = await vi.importActual<
+      typeof import("../../../domains/singles/services/reconcile-single-companion")
+    >("../../../domains/singles/services/reconcile-single-companion");
+    return {
+      ...actual,
+      reconcileSingleCompanion: vi.fn(
+        async (args: Parameters<typeof actual.reconcileSingleCompanion>[0]) => {
+          const result = await actual.reconcileSingleCompanion(args);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/dispatcher/handlers/__tests__/single-dispatcher-update-ddl.test.ts:159]: Confirmed exactly as described, and the earlier thread saying it was fixed was wrong. Fixed in `64e230a21`.

The scripted edit that was supposed to add the `before` branch matched nothing — the anchor had been reformatted by prettier at commit time — so `companionFailure.when` was written and never read, and both rows exercised the post-DDL state. I claimed it fixed without re-reading the file. That is the second time in this PR a patch of mine silently failed to apply, and I have switched to edits that error when the anchor is absent rather than replacements that no-op.

The branch is in place now, before the real call. More importantly, the parameterization no longer depends on my having wired it correctly:

```ts
expect(realReconcileCalls.count).toBe(when === "before" ? 0 : 1);
```

The two rows must reach DIFFERENT states — "before" must not have run the reconcile at all, "after" must have run it to completion — so a fixture that quietly collapses them fails instead of passing. That is the assertion that was missing, and its absence is exactly why the bug survived a round.

Restoring the defect you found (deleting the `before` branch so the real call is unconditional) now fails that row alone, 1 failed / 12 passed.

To your point about what it was hiding: the recovery from a rejected first companion statement was in fact correct — retraction is indifferent to where the reconcile stopped, which is the property that made it the right answer. But that was luck rather than evidence until now, since nothing was testing it.

**Comment by mobeenabdullah** [comment]
@codex please review this PR
